# Common table expressions

Using the northwind database

In [1]:
import pymysql
import sqlalchemy as alch
from getpass import getpass
import pandas as pd
import os

## 1. Write a CTE that lists the names and quantities of products with a unit price greater than $50.

In [2]:
from getpass import getpass
password = getpass()
dbName = "northwind"
connectionData=f"mysql+pymysql://root:{password}@localhost/{dbName}"
engine = alch.create_engine(connectionData)

```sql
-- Your code down below

```

In [3]:
query = """WITH set_products (prod_name, p_unit)
                AS 
        (SELECT p.ProductName,p.Unit  FROM Products p
         WHERE p.Price > 50
        )
        SELECT prod_name, p_unit 
            FROM set_products; 
"""
df = pd.read_sql_query(query, engine)
df

prod_name                p_unit
0          Mishi Kobe Niku      18 - 500 g pkgs.
1         Carnarvon Tigers            16 kg pkg.
2   Sir Rodney's Marmalade         30 gift boxes
3  Thüringer Rostbratwurst  50 bags x 30 sausgs.
4            Côte de Blaye    12 - 75 cl bottles
5    Manjimup Dried Apples      50 - 300 g pkgs.
6     Raclette Courdavault             5 kg pkg.

#### Expected results

|ProductName                     |Unit|
|--------------------------------|----|
|Mishi Kobe Niku                 |18 - 500 g pkgs.|
|Carnarvon Tigers                |16 kg pkg.|
|Sir Rodney's Marmalade          |30 gift boxes|
|Thüringer Rostbratwurst         |50 bags x 30 sausgs.|
|Côte de Blaye                   |12 - 75 cl bottles|
|Manjimup Dried Apples           |50 - 300 g pkgs.|
|Raclette Courdavault            |5 kg pkg.|


## 2. What are the top 5 most profitable products?

```sql
-- Your code down below
```

In [11]:
query = """WITH top_profit (ProductID, ProductName,TotalRevenue)
                AS 
        (SELECT p.ProductID ,p.ProductName,(sum(od.Quantity)*p.Price) AS TotalRevenue FROM Products p 
JOIN OrderDetails od ON od.ProductID = p.ProductID  
JOIN Orders o ON o.OrderID = od.OrderID 
GROUP BY p.ProductID , p.ProductName 
ORDER BY TotalRevenue desc
        )
        SELECT ProductID, ProductName,TotalRevenue 
            FROM top_profit limit 5
"""
df = pd.read_sql_query(query, engine)
df

ProductID              ProductName  TotalRevenue
0         38            Côte de Blaye       63096.0
1         29  Thüringer Rostbratwurst       20832.0
2         59     Raclette Courdavault       19030.0
3         62           Tarte au sucre       15925.0
4         60        Camembert Pierrot       14620.0

#### Expected results
Alice Mutton is 6th

|ProductID                       |ProductName|TotalRevenue|
|--------------------------------|-----------|------------|
|-                             |-|-       |
|-                           |-|-       |
|-                            |-|-       |
|-                             |-|-       |
|-                              |-|-       |
|17                              |Alice Mutton|12909       |

## 3. Write a CTE that lists the top 5 categories by the number of products they have.

```sql
-- Your code down below
```

In [15]:
query = ''' 
WITH top_cats (CategoryName,ProductCount) AS
(
SELECT c.CategoryName ,count(p.ProductName) AS total_products  FROM Categories c 
JOIN Products p ON p.CategoryID = c.CategoryID 
GROUP BY c.CategoryName
ORDER BY total_products DESC
)
SELECT CategoryName, ProductCount FROM top_cats limit 5
'''

df = pd.read_sql_query(query, engine)
df

CategoryName  ProductCount
0     Confections            13
1       Beverages            12
2      Condiments            12
3         Seafood            12
4  Dairy Products            10

#### Expected result
|CategoryName                    |ProductCount|
|--------------------------------|------------|
|Confections                     |13          |
|Beverages                       |12          |
|Condiments                      |12          |
|Seafood                         |12          |
|Dairy Products                  |10          |


## 4. Write a CTE that shows the average order quantity for each product category.

```sql
-- Your code down below

```

In [17]:
query = ''' 
WITH avg_cat (CategoryName,CatCount,QtySum) AS
(
SELECT c.CategoryName,count(od.OrderID),sum(od.Quantity)  FROM Products p 
JOIN OrderDetails od  ON od.ProductID = p.ProductID 
JOIN Categories c ON c.CategoryID = p.CategoryID
GROUP BY c.CategoryName 
)
SELECT CategoryName, QtySum/CatCount as AvgOrderQuantity FROM avg_cat
'''

df = pd.read_sql_query(query, engine)
df

CategoryName  AvgOrderQuantity
0       Beverages           24.6129
1      Condiments           28.2245
2     Confections           25.1190
3  Dairy Products           26.0100
4  Grains/Cereals           21.7143
5    Meat/Poultry           25.7600
6         Produce           21.6667
7         Seafood           21.5672

|CategoryName                    |AvgOrderQuantity|
|--------------------------------|----------------|
|Beverages                       |24.6129         |
|Condiments                      |28.2245         |
|Confections                     |25.1190         |
|Dairy Products                  |26.0100         |
|Grains/Cereals                  |21.7143         |
|Meat/Poultry                    |25.7600         |
|Produce                         |21.6667         |
|Seafood                         |21.5672         |


# 5. Create a CTE to calculate the average order amount for each customer.

```sql
-- Your code down below


```

In [44]:
query = ''' 
WITH avg_customer (CustomerID,CustomerName,TotAmount,TotOrders) AS
(
SELECT c.CustomerID ,c.CustomerName , sum(p.Price*od.Quantity), count(o.OrderID) AS c_orders FROM Customers c 
JOIN Orders o ON o.CustomerID = c.CustomerID 
JOIN OrderDetails od ON od.OrderID = o.OrderID 
JOIN Products p ON p.ProductID = od.ProductID 
GROUP BY c.CustomerID ,c.CustomerName
ORDER BY c_orders DESC
)
SELECT CustomerID, CustomerName, TotAmount/Totorders as AvgOrderAmount FROM avg_customer
order by AvgOrderAmount desc
'''

df = pd.read_sql_query(query, engine)
df

CustomerID                        CustomerName  AvgOrderAmount
0           59                    Piccolo und mehr       4014.2500
1           73                       Simons bistro       2444.3333
2           62                       Queen Cozinha       1991.6667
3           51                      Mère Paillarde       1673.8571
4           71                  Save-a-lot Markets       1407.2500
..         ...                                 ...             ...
69          29              Galería del gastrónomo        147.7500
70          69                    Romero y tomillo        116.5000
71          13          Centro comercial Moctezuma         63.0000
72          27                      Franchi S.p.A.         62.0000
73           2  Ana Trujillo Emparedados y helados         55.5000

[74 rows x 3 columns]

|CustomerID                      |CustomerName|AvgOrderAmount|
|--------------------------------|------------|--------------|
|59                              |Piccolo und mehr|4014.2500     |
|73                              |Simons bistro|2444.3333     |
|62                              |Queen Cozinha|1991.6667     |
|51                              |Mère Paillarde|1673.8571     |
|71                              |Save-a-lot Markets|1407.2500     |
|76                              |Suprêmes délices|1345.8333     |
|81                              |Tradição Hipermercados|1315.6667     |
|7                               |Blondel père et fils|1174.4615     |
|89                              |White Clover Markets|1112.5000     |
|55                              |Old World Delicatessen|1079.5000     |
|20                              |Ernst Handel|1018.0000     |
|19                              |Eastern Connection|1004.8000     |
|68                              |Richter Supermarkt|976.6667      |
|72                              |Seven Seas Imports|934.0000      |
|63                              |QUICK-Stop  |908.5500      |
|25                              |Frankenversand|895.0667      |
|75                              |Split Rail Beer & Ale|854.7692      |
|65                              |Rattlesnake Canyon Grocery|838.2727      |
|52                              |Morgenstern Gesundkost|754.0000      |
|9                               |Bon app''   |750.4286      |
|37                              |Hungry Owl All-Night Grocers|733.5714      |
|35                              |HILARIÓN-Abastos|722.8333      |
|33                              |GROSELLA-Restaurante|690.0000      |
|34                              |Hanari Carnes|681.0000      |
|15                              |Comércio Mineiro|677.5000      |
|23                              |Folies gourmandes|672.3333      |
|10                              |Bottom-Dollar Marketse|664.0000      |
|31                              |Gourmet Lanchonetes|641.0000      |
|60                              |Princesa Isabel Vinhoss|628.5000      |
|5                               |Berglunds snabbköp|601.5556      |
|46                              |LILA-Supermercado|564.7692      |
|44                              |Lehmanns Marktstand|545.7500      |
|30                              |Godos Cocina Típica|515.0000      |
|3                               |Antonio Moreno Taquería|504.0000      |
|88                              |Wellington Importadora|503.8333      |
|47                              |LINO-Delicateses|500.0000      |
|56                              |Ottilies Käseladen|500.0000      |
|87                              |Wartian Herkku|492.5833      |
|24                              |Folk och fä HB|480.2222      |
|49                              |Magazzini Alimentari Riuniti|464.2857      |
|86                              |Die Wandernde Kuh|441.9091      |
|41                              |La maison d''Asie|437.8182      |
|8                               |Bólido Comidas preparadas|416.0000      |
|83                              |Vaffeljernet|411.0000      |
|14                              |Chop-suey Chinese|374.8333      |
|80                              |Tortuga Restaurante|357.4167      |


# 6. Sales Analysis with CTEs

Assume we have the Northwind database which contains tables like Orders, OrderDetails, and Products. Create a CTE that calculates the total sales for each product in the year 1997.

```sql
-- Your code down below

 ```

In [37]:
query = ''' 
WITH total_1997 (ProductName,TotalSales) AS
(
SELECT p.ProductName,sum(od.Quantity) FROM OrderDetails od
JOIN Orders o ON o.OrderID  = od.OrderID 
JOIN Products p ON p.ProductID = od.ProductID 
WHERE YEAR(o.OrderDate) = 1997
GROUP BY p.ProductName
)
SELECT ProductName,TotalSales from total_1997
order by TotalSales desc
limit 15

'''

df = pd.read_sql_query(query, engine)
df


ProductName  TotalSales
0   Gnocchi di nonna Alice       173.0
1                Tourtière       126.0
2                  Geitost       119.0
3     Raclette Courdavault       115.0
4                    Chang       115.0
5           Sirop d'érable       106.0
6             Vegie-spread       100.0
7            Côte de Blaye        99.0
8             Alice Mutton        97.0
9           Steeleye Stout        95.0
10     Sir Rodney's Scones        92.0
11                 Pavlova        86.0
12           Zaanse koeken        85.0
13             Fløtemysost        75.0
14          Tarte au sucre        75.0

#### Expected result

|ProductName                     |TotalSales|
|--------------------------------|----------|
|Gnocchi di nonna Alice          |173       |
|Tourtière                       |126       |
|Geitost                         |119       |
|Chang                           |115       |
|Raclette Courdavault            |115       |
|Sirop d'érable                  |106       |
|Vegie-spread                    |100       |
|Côte de Blaye                   |99        |
|Alice Mutton                    |97        |
|Steeleye Stout                  |95        |
|Sir Rodney's Scones             |92        |
|Pavlova                         |86        |
|Zaanse koeken                   |85        |
|Fløtemysost                     |75        |
|Tarte au sucre                  |75        |
